pandevice/objects.py

```diff
@@ -147,3 +147,256 @@ def color(self):
+        description (str): Description of this object
+        tag (list): Administrative tags
+    """
+    ROOT = Root.VSYS
+    SUFFIX = ENTRY
+
+    def _setup(self):
+        # xpaths
+        self._xpaths.add_profile(value='/service')
+
+        # params
+        params = []
+
+        params.append(VersionedParamPath(
+            'protocol', path='protocol/{protocol}',
+            values=['tcp', 'udp'], default='tcp'))
+        params.append(VersionedParamPath(
+            'source_port', path='protocol/{protocol}/source-port'))
+        params.append(VersionedParamPath(
+            'port', path='protocol/{protocol}/port', default='1'))
+        params.append(VersionedParamPath(
+            'description', path='description'))
+        params.append(VersionedParamPath(
+            'tag', path='tag', vartype='member'))
+
+        self._params = tuple(params)
+
+
+class ServiceGroup(VersionedPanObject):
+    """ServiceGroup Object
+
+    Args:
+        name (str): Name of the object
+        value (list): List of service values
+        tag (list): Administrative tags
+    """
+    ROOT = Root.VSYS
+    SUFFIX = ENTRY
+
+    def _setup(self):
+        # xpaths
+        self._xpaths.add_profile(value='/service-group')
+
+        # params
+        params = []
+
+        params.append(VersionedParamPath(
+            'value', path='members', vartype='member'))
+        params.append(VersionedParamPath(
+            'tag', path='tag', vartype='member'))
+
+        self._params = tuple(params)
+
+
+class ApplicationObject(VersionedPanObject):
+    """Application Object
+
+    Args:
+        name (str): Name of the object
+        category (str): Application category
+        subcategory (str): Application subcategory
+        technology (str): Application technology
+        risk (str): 
+        default_type (str): 
+        default_value (list): 
+        parent_app (str): 
+        timeout (str): 
+        tcp_timeout (str): 
+        udp_timeout (str): 
+        tcp_half_closed_timeout (str): 
+        tcp_time_wait_timeout (str): 
+        evasive_behavior (bool): 
+        consume_big_bandwidth (bool): 
+        used_by_malware (bool): 
+        able_to_transfer_file (bool): 
+        has_known_vulnerability (bool): 
+        tunnel_other_application (bool): 
+        tunnel_applications (bool): 
+        prone_to_misuse (bool): 
+        pervasive_use (bool): 
+        file_type_ident (bool): 
+        virus_ident (bool): 
+        data_ident (bool): 
+        description (str): Description of this object
+        tag (list): Administrative tags
+    
+    Please refer to https://applipedia.paloaltonetworks.com/ for more info on these params
+    """
+    ROOT = Root.VSYS
+    SUFFIX = ENTRY
+
+    def _setup(self):
+        # xpaths
+        self._xpaths.add_profile(value='/application')
+
+        # TODO: default_types other than 'port' are not "member" params, need to figure out what to do with those
```

Comment:
Regarding your 'TODO' item on line 255, this can be done with a `conditional` on a parameter, but currently we only support 'and' conditionals, and this would require an 'or' conditional, so I'll have to make a small adjustment to `base.py`.  I just tested the adjustment and will add it to the 'develop' branch after this pull request is merged, so don't worry about it for this pull request.  No changes needed here, this comment is just for your information.